In [2]:
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
import joblib
import pandas as pd


In [3]:
import gdown
import pandas as pd

# Define the URL of the CSV file
file_id = '1WClX4TtAVny-bz7AI1VNrqeGlnzyXiam'
url = f'https://drive.google.com/uc?id={file_id}'

# Download the file using gdown
output = 'sampled_data_full.csv'
gdown.download(url, output, quiet=False)

# Read the CSV file into a DataFrame
df = pd.read_csv(output)

# Display the last few rows of the DataFrame
df.tail()

Downloading...
From (original): https://drive.google.com/uc?id=1WClX4TtAVny-bz7AI1VNrqeGlnzyXiam
From (redirected): https://drive.google.com/uc?id=1WClX4TtAVny-bz7AI1VNrqeGlnzyXiam&confirm=t&uuid=d386b0e5-9c68-471d-bc21-c4d97624be03
To: /Users/jinluwang/Downloads/Do work here/Challenges/predictive_flight_delays/notebooks/sampled_data_full.csv
100%|██████████████████████████████████████████████| 268M/268M [00:06<00:00, 38.9MB/s]


MONTH  DAY_OF_WEEK  DEP_DEL15 DEP_TIME_BLK  DISTANCE_GROUP  \
1297807      6            5          0    0800-0859               7   
1297808     10            3          0    1100-1159               4   
1297809      5            3          0    1900-1959               2   
1297810      6            3          1    0700-0759               4   
1297811      6            2          0    0700-0759               3   

         SEGMENT_NUMBER  CONCURRENT_FLIGHTS  NUMBER_OF_SEATS  \
1297807               1                  37              150   
1297808               3                   5               76   
1297809               5                   9               70   
1297810               2                  45               76   
1297811               1                  13              173   

                  CARRIER_NAME  AIRPORT_FLIGHTS_MONTH  ...  PLANE_AGE  \
1297807        JetBlue Airways                  12427  ...          8   
1297808  SkyWest Airlines Inc.                   5961  ...          0   
1297809            Comair Inc.                   2558  ...         15   
1297810  SkyWest Airlines Inc.                  21878  ...          4   
1297811  United Air Lines Inc.                   4615  ...          6   

                       DEPARTING_AIRPORT  LATITUDE  LONGITUDE  \
1297807              Logan International    42.364    -71.006   
1297808  Washington Dulles International    38.942    -77.458   
1297809            Bradley International    41.939    -72.683   
1297810          Stapleton International    39.774   -104.880   
1297811        Kansas City International    39.298    -94.714   

                          PREVIOUS_AIRPORT  PRCP  SNOW SNWD  TMAX   AWND  
1297807                               NONE   0.0   0.0  0.0  77.0  14.32  
1297808         Detroit Metro Wayne County   0.0   0.0  0.0  96.0   6.04  
1297809  Ronald Reagan Washington National   0.0   0.0  0.0  68.0   6.93  
1297810       Billings Logan International   0.0   0.0  0.0  83.0   7.83  
1297811                               NONE   0.0   0.0  0.0  84.0   9.40  

[5 rows x 26 columns]

In [4]:
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

In [5]:
# Sample 20% of the data
data_sample = df.sample(frac=0.2, random_state=42)

# Define features and target
features = ['MONTH', 'DAY_OF_WEEK', 'DEP_TIME_BLK',
            'TMAX', 'AWND','AIRPORT_FLIGHTS_MONTH','AIRLINE_FLIGHTS_MONTH','AIRLINE_AIRPORT_FLIGHTS_MONTH','PRCP']
print(len(features))
# Prepare the data
X = data_sample[features]
y = data_sample['DEP_DEL15']

9


In [6]:
data_sample[features]

MONTH  DAY_OF_WEEK DEP_TIME_BLK  TMAX   AWND  AIRPORT_FLIGHTS_MONTH  \
738784       8            4    1300-1359  92.0   3.13                  15106   
1222149      5            5    0800-0859  65.0   5.37                  12849   
63596        1            4    1600-1659  44.0   4.70                   6023   
1183182      7            6    0700-0759  89.0   7.83                   1598   
1003874      3            7    1400-1459  62.0  14.32                   1375   
...        ...          ...          ...   ...    ...                    ...   
918134      10            7    1000-1059  88.0  10.07                  25833   
772057       2            5    0600-0659  21.0   5.14                  22752   
51002       10            2    2000-2059  79.0   8.72                  20449   
375745       5            4    0800-0859  83.0  10.51                  12037   
207473       7            6    0900-0959  72.0   6.04                  13781   

         AIRLINE_FLIGHTS_MONTH  AIRLINE_AIRPORT_FLIGHTS_MONTH  PRCP  
738784                   22671                           1061  0.00  
1222149                 113709                            948  0.24  
63596                    62105                            559  0.00  
1183182                  72721                            731  0.00  
1003874                 114119                            535  0.00  
...                        ...                            ...   ...  
918134                   71625                           1405  0.97  
772057                   43512                           5161  0.02  
51002                    24198                           8189  0.57  
375745                   22451                           5958  0.00  
207473                   24454                           6465  0.00  

[259562 rows x 9 columns]

In [7]:
X.columns

Index(['MONTH', 'DAY_OF_WEEK', 'DEP_TIME_BLK', 'TMAX', 'AWND',
       'AIRPORT_FLIGHTS_MONTH', 'AIRLINE_FLIGHTS_MONTH',
       'AIRLINE_AIRPORT_FLIGHTS_MONTH', 'PRCP'],
      dtype='object')

In [8]:
X.iloc[0]

MONTH                                    8
DAY_OF_WEEK                              4
DEP_TIME_BLK                     1300-1359
TMAX                                  92.0
AWND                                  3.13
AIRPORT_FLIGHTS_MONTH                15106
AIRLINE_FLIGHTS_MONTH                22671
AIRLINE_AIRPORT_FLIGHTS_MONTH         1061
PRCP                                   0.0
Name: 738784, dtype: object

In [9]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Index: 259562 entries, 738784 to 207473
Data columns (total 9 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   MONTH                          259562 non-null  int64  
 1   DAY_OF_WEEK                    259562 non-null  int64  
 2   DEP_TIME_BLK                   259562 non-null  object 
 3   TMAX                           259562 non-null  float64
 4   AWND                           259562 non-null  float64
 5   AIRPORT_FLIGHTS_MONTH          259562 non-null  int64  
 6   AIRLINE_FLIGHTS_MONTH          259562 non-null  int64  
 7   AIRLINE_AIRPORT_FLIGHTS_MONTH  259562 non-null  int64  
 8   PRCP                           259562 non-null  float64
dtypes: float64(3), int64(5), object(1)
memory usage: 19.8+ MB


In [10]:
label_enc = LabelEncoder()
encoded_time = label_enc.fit_transform(X['DEP_TIME_BLK'])
joblib.dump(label_enc, "../trained_modules/label_encoder.pkl")

['../trained_modules/label_encoder.pkl']

In [11]:
X =X.drop(columns=["DEP_TIME_BLK"])

In [12]:
X["ENCODED_DEP_TIME_BLK"] = encoded_time

In [13]:
X.isna().sum()

MONTH                            0
DAY_OF_WEEK                      0
TMAX                             0
AWND                             0
AIRPORT_FLIGHTS_MONTH            0
AIRLINE_FLIGHTS_MONTH            0
AIRLINE_AIRPORT_FLIGHTS_MONTH    0
PRCP                             0
ENCODED_DEP_TIME_BLK             0
dtype: int64

In [14]:
X.columns

Index(['MONTH', 'DAY_OF_WEEK', 'TMAX', 'AWND', 'AIRPORT_FLIGHTS_MONTH',
       'AIRLINE_FLIGHTS_MONTH', 'AIRLINE_AIRPORT_FLIGHTS_MONTH', 'PRCP',
       'ENCODED_DEP_TIME_BLK'],
      dtype='object')

In [15]:
# Handle categorical variables
# categorical_features = ['CARRIER_NAME', 'DEPARTING_AIRPORT', 'PREVIOUS_AIRPORT']
# X = pd.get_dummies(X, columns=categorical_features, drop_first=True)

# Handle 'DEP_TIME_BLK' separately
# X['DEP_TIME_BLK'] = X['DEP_TIME_BLK'].astype('category').cat.codes

# Make sure all columns are numeric
# X = X.apply(pd.to_numeric, errors='coerce')

# Drop any rows with NaN values
# X.dropna(inplace=True)
# y = y[X.index]

In [16]:
# Use pandas for categorical to numerical conversion.

In [17]:
# Normalize the data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


In [18]:
joblib.dump(scaler, "../trained_modules/standard_scaler.pkl")

['../trained_modules/standard_scaler.pkl']

In [19]:
scaler.get_feature_names_out()

array(['MONTH', 'DAY_OF_WEEK', 'TMAX', 'AWND', 'AIRPORT_FLIGHTS_MONTH',
       'AIRLINE_FLIGHTS_MONTH', 'AIRLINE_AIRPORT_FLIGHTS_MONTH', 'PRCP',
       'ENCODED_DEP_TIME_BLK'], dtype=object)

In [20]:
# Split the data
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

In [21]:
# Build the KNN model
knn_model = KNeighborsClassifier(n_neighbors=5)  # You can tune the n_neighbors parameter

# Train the model
knn_model.fit(X_train, y_train)


KNeighborsClassifier()

In [22]:
# Evaluate the model
y_pred = knn_model.predict(X_test)
test_acc = knn_model.score(X_test, y_test)
print('\nTest accuracy:', test_acc)

# Save the model
joblib.dump(knn_model, "../trained_modules/knn_model.pkl")

from sklearn.metrics import confusion_matrix, classification_report
# Print confusion matrix and classification report
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print('\nConfusion Matrix:')
print(conf_matrix)

print('\nClassification Report:')
print(class_report)


Test accuracy: 0.7966983221928997

Confusion Matrix:
[[39722  2564]
 [ 7990  1637]]

Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.94      0.88     42286
           1       0.39      0.17      0.24      9627

    accuracy                           0.80     51913
   macro avg       0.61      0.55      0.56     51913
weighted avg       0.75      0.80      0.76     51913



In [24]:
# Evaluate the model
y_pred = knn_model.predict(X_test)
test_acc = knn_model.score(X_test, y_test)
print('\nTest accuracy:', test_acc)

# Save the model
joblib.dump(knn_model, "../trained_modules/knn_model_2.pkl",compress=3)


Test accuracy: 0.7966983221928997


['../trained_modules/knn_model_2.pkl']

In [25]:
from sklearn.metrics import confusion_matrix, classification_report
# Print confusion matrix and classification report
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print('\nConfusion Matrix:')
print(conf_matrix)

print('\nClassification Report:')
print(class_report)


Confusion Matrix:
[[39722  2564]
 [ 7990  1637]]

Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.94      0.88     42286
           1       0.39      0.17      0.24      9627

    accuracy                           0.80     51913
   macro avg       0.61      0.55      0.56     51913
weighted avg       0.75      0.80      0.76     51913



In [ ]:
from sklearn.metrics import confusion_matrix, classification_report
# Print confusion matrix and classification report
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print('\nConfusion Matrix:')
print(conf_matrix)

print('\nClassification Report:')
print(class_report)

Compare to the NNW model               precision    recall  f1-score   support

           0       0.82      0.99      0.90     42286
           1       0.54      0.06      0.11      9627

    accuracy                           0.82     51913
   macro avg       0.68      0.52      0.50     51913
weighted avg       0.77      0.82      0.75     51913


Key Metrics to Consider
Precision: The ratio of true positive predictions to the total predicted positives.
Recall: The ratio of true positive predictions to the total actual positives.
F1-score: The harmonic mean of precision and recall, providing a single metric that balances both concerns.
Support: The number of actual occurrences of the class in the dataset.
Detailed Comparison
Accuracy
Report 1: 0.82
Report 2: 0.80
Observation: Report 1 has a slightly higher overall accuracy.
Precision for Class 1
Report 1: 0.54
Report 2: 0.39
Observation: Report 1 is better at correctly identifying delays when it predicts a delay.
Recall for Class 1
Report 1: 0.06
Report 2: 0.17
Observation: Report 2 is better at finding the actual delays, although both are quite low.
F1-score for Class 1
Report 1: 0.11
Report 2: 0.24
Observation: Report 2 has a higher F1-score, indicating a better balance between precision and recall for class 1 (delays).
Weighted Average F1-score
Report 1: 0.75
Report 2: 0.76
Observation: Both are quite similar, with Report 2 being marginally better.
Conclusion
Report 2 appears to be the better choice overall. Although the accuracy and precision for class 1 are slightly lower than in Report 1, the recall and F1-score for class 1 are significantly higher. This suggests that Report 2 is better at identifying true delays, which is often more important in imbalanced datasets like this one.

The weighted average F1-score is also slightly higher in Report 2, indicating a better overall balance across both classes. Therefore, despite the lower recall and precision for class 1, Report 2 provides a better balance and performance for identifying flight delays.